src/pymor/algorithms/timestepping.py

```diff
@@ -10,227 +10,609 @@
 by an external library into an instationary |Model|.
 
 Currently, implementations of :func:`explicit_euler` and :func:`implicit_euler`
-time-stepping are provided. The :class:`TimeStepper` defines a
-common interface that has to be fulfilled by the time-steppers used
-by |InstationaryModel|. The classes :class:`ExplicitEulerTimeStepper`
-and :class:`ImplicitEulerTimeStepper` encapsulate :func:`explicit_euler` and
-:func:`implicit_euler` to provide this interface.
+time-stepping are provided, based on the :class:`ExplicitEulerTimeStepper` and
+:class:`ImplicitEulerTimeStepper` classes. These derive from :class:`TimeStepper`,
+which defines a common interface that has to be fulfilled by the time-steppers used
+by |InstationaryModel|.
 """
 
-from pymor.core.base import ImmutableObject, abstractmethod
+from numbers import Number
+import numpy as np
+
+from pymor.core.base import BasicObject, ImmutableObject, abstractmethod
+from pymor.operators.constructions import IdentityOperator, VectorArrayOperator, ZeroOperator
 from pymor.operators.interface import Operator
+from pymor.parameters.base import Mu
+from pymor.tools import floatcmp
+from pymor.tools.deprecated import Deprecated
 from pymor.vectorarrays.interface import VectorArray
 
 
 class TimeStepper(ImmutableObject):
     """Interface for time-stepping algorithms.
 
-    Algorithms implementing this interface solve time-dependent problems
+    Algorithms implementing this interface solve time-dependent initial value problems
     of the form ::
 
-        M * d_t u + A(u, mu, t) = F(mu, t).
+        M * d_t u + A(u, mu, t) = F(mu, t),
+                     u(mu, t_0) = u_0(mu).
+
+    Time-steppers used by |InstationaryModel| have to fulfill this interface. Time evolution can be performed
+    by calling :meth:`solve`.
 
-    Time-steppers used by |InstationaryModel| have to fulfill
-    this interface.
+    Note that the actual work is done in an iterator derived from :class:`TimeStepperIterator`.
+
+    Parameters
+    ----------
+    num_values
+        The number of returned vectors of the solution trajectory. If `None`, each intermediate vector that is
+        calculated is returned. Else an interpolation of the calculated vectors on an equidistant temporal grid is
+        returned, using an appropriate interpolation of the respective time stepper.
     """
 
-    @abstractmethod
-    def solve(self, initial_time, end_time, initial_data, operator, rhs=None, mass=None, mu=None, num_values=None):
+    IteratorType = None
+    available_interpolations = ('P0', 'P1')
+
+    def __init__(self, num_values=None, interpolation='P1'):
+        assert not num_values or (isinstance(num_values, Number) and num_values > 1)
+        assert interpolation in self.available_interpolations
+        self.__auto_init(locals())
+
+    def solve(self, t0, t1, U0, A, F=None, M=None, mu=None, iter=False, return_times=False):
```

Comment:
`iter` clashes with a builtin, could we rename this?

Reply:
we decided to name it like this on purpose, if I recall it correctly @sdrave?

Reply:
renamed to `return_iter`